File: src/main/java/net/pterodactylus/sone/web/EditProfileFieldPage.java
/*
 * Sone - EditProfileFieldPage.java - Copyright © 2011 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.web;

import net.pterodactylus.sone.data.Profile;
import net.pterodactylus.sone.data.Profile.Field;
import net.pterodactylus.sone.data.Sone;
import net.pterodactylus.sone.web.page.Page.Request.Method;
import net.pterodactylus.util.template.Template;
import net.pterodactylus.util.template.TemplateContext;

/**
 * Page that lets the user edit the name of a profile field.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class EditProfileFieldPage extends SoneTemplatePage {

	/**
	 * Creates a new “edit profile field” page.
	 *
	 * @param template
	 *            The template to render
	 * @param webInterface
	 *            The Sone web interface
	 */
	public EditProfileFieldPage(Template template, WebInterface webInterface) {
		super("editProfileField.html", template, "Page.EditProfileField.Title", webInterface, true);
	}

	//
	// SONETEMPLATEPAGE METHODS
	//

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected void processTemplate(Request request, TemplateContext templateContext) throws RedirectException {
		super.processTemplate(request, templateContext);
		Sone currentSone = getCurrentSone(request.getToadletContext());
		Profile profile = currentSone.getProfile();

		/* get parameters from request. */
		String fieldId = request.getHttpRequest().getParam("field");
		Field field = profile.getFieldById(fieldId);
		if (field == null) {
			throw new RedirectException("invalid.html");
		}

		/* process the POST request. */
		if (request.getMethod() == Method.POST) {
			if (request.getHttpRequest().getPartAsStringFailsafe("cancel", 4).equals("true")) {
				throw new RedirectException("editProfile.html#profile-fields");
			}
			fieldId = request.getHttpRequest().getPartAsStringFailsafe("field", 36);
			field = profile.getFieldById(fieldId);
			if (field == null) {
				throw new RedirectException("invalid.html");
			}
			String name = request.getHttpRequest().getPartAsStringFailsafe("name", 256);
			Field existingField = profile.getFieldByName(name);
			if ((existingField == null) || (existingField.equals(field))) {
				field.setName(name);
				currentSone.setProfile(profile);
				throw new RedirectException("editProfile.html#profile-fields");
			}
			templateContext.set("duplicateFieldName", true);
		}

		/* store current values in template. */
		templateContext.set("field", field);
	}

}

File: src/main/kotlin/net/pterodactylus/sone/web/pages/EditProfileFieldPage.kt
package net.pterodactylus.sone.web.pages

import net.pterodactylus.sone.data.*
import net.pterodactylus.sone.main.*
import net.pterodactylus.sone.utils.*
import net.pterodactylus.sone.web.*
import net.pterodactylus.sone.web.page.*
import net.pterodactylus.util.template.*
import javax.inject.*

/**
 * Page that lets the user edit the name of a profile field.
 */
@TemplatePath("/templates/editProfileField.html")
@ToadletPath("editProfileField.html")
class EditProfileFieldPage @Inject constructor(webInterface: WebInterface, loaders: Loaders, templateRenderer: TemplateRenderer) :
		LoggedInPage("Page.EditProfileField.Title", webInterface, loaders, templateRenderer) {

	override fun handleRequest(soneRequest: SoneRequest, currentSone: Sone, templateContext: TemplateContext) {
		currentSone.profile.let { profile ->
			if (soneRequest.isPOST) {
				if (soneRequest.httpRequest.getPartAsStringFailsafe("cancel", 4) == "true") {
					redirectTo("editProfile.html#profile-fields")
				}
				val field = profile.getFieldById(soneRequest.httpRequest.getPartAsStringFailsafe("field", 36)) ?: redirectTo("invalid.html")
				soneRequest.httpRequest.getPartAsStringFailsafe("name", 256).let { name ->
					try {
						if (name != field.name) {
							field.name = name
							currentSone.profile = profile
						}
						redirectTo("editProfile.html#profile-fields")
					} catch (e: IllegalArgumentException) {
						templateContext["duplicateFieldName"] = true
						return
					}
				}
			}
			templateContext["field"] = profile.getFieldById(soneRequest.httpRequest.getParam("field")) ?: redirectTo("invalid.html")
		}
	}

}
